spec/integration/create_subscription_spec.rb

```diff
@@ -0,0 +1,172 @@
+RSpec.describe "Creating a subscription", type: :request do
+  let!(:subscriber_list) { create(:subscriber_list) }
+  let!(:subscriber) { create(:subscriber, address: "test@example.com") }
+  let(:address) { subscriber.address }
+  let(:frequency) { "daily" }
+
+  def create_subscription
+    post "/subscriptions", params: { subscriber_list_id: subscriber_list.id, address: address, frequency: frequency }
+  end
+
+  context "with authentication and authorisation" do
+    before do
+      login_with_internal_app
+    end
+
+    it "requires an address parameter" do
+      post "/subscriptions", params: { subscriber_list_id: 10 }
+      expect(response.status).to eq(400)
+    end
+
+    it "requires a subscriber_list_id parameter" do
+      post "/subscriptions", params: { address: "test@example.com" }
+      expect(response.status).to eq(400)
+    end
+
+    it "returns status code 404 if the subscriber list doesn't exist" do
+      post "/subscriptions", params: { subscriber_list_id: 10, address: "test@example.com" }
+      expect(response.status).to eq(404)
+    end
+
+    context "with an existing subscription with different frequency" do
+      let!(:subscription) { create(:subscription, subscriber_list: subscriber_list, subscriber: subscriber, frequency: :immediately) }
+
+      it "creates a new subscription" do
+        expect { create_subscription }.to change(Subscription, :count)
+      end
+
+      it "returns status code 200" do
+        create_subscription
+        expect(response.status).to eq(200)
+      end
+
+      it "returns the ID of the new subscription" do
+        create_subscription
+        expect(data[:id]).to_not eq(subscription.id)
+      end
+
+      it "marks the existing subscription as ended" do
+        create_subscription
+        expect(subscription.reload.ended?).to be true
+      end
+
+      it "sends a confirmation email" do
+        stub_notify
+        create_subscription
+        expect(a_request(:post, /notifications/)).to have_been_made.at_least_once
+      end
+    end
+
+    context "with a parameter to skip sending the confirmation email" do
+      def create_subscription
```

Comment:
I wondered whether this should be named differently from the other `create_subscription`? It applies to quite a small block at the moment though, so I'm not too worried.

Reply:
Oh, how did it read for you? I deliberately named it the same to indicate it should be used in preference to the outer one, but looking again I also see a bit of duplication going on.

I've added another commit with a different approach here - could you take a look?

Reply:
Yep - I think I prefer the way you went in the new commit